substrate/client/sysinfo/src/sysinfo.rs

```diff
@@ -375,6 +379,49 @@ pub fn benchmark_cpu(limit: ExecutionLimit) -> Throughput {
 		.expect("benchmark cannot fail; qed")
 }
 
+// This benchmarks the entire CPU speed as measured by calculating BLAKE2b-256 hashes, in bytes per
+// second. It spawns multiple threads to measure the throughput of the entire CPU and averages the
+// score obtained by each thread. If we have at least `EXPECTED_NUM_CORES` available then the
+// average throughput should be relatively close to the single core performance as measured by
+// `benchmark_cpu`.
+pub fn benchmark_cpu_parallelism(limit: ExecutionLimit) -> Throughput {
+	const SIZE: usize = 32 * 1024;
+	const EXPECTED_NUM_CORES: usize = 8;
+
+	let ready_to_run_benchmark = Arc::new(Barrier::new(EXPECTED_NUM_CORES));
+	let mut benchmark_threads = Vec::new();
+
+	// Spawn a thread for each expected core and average the throughput for each of them.
+	for _ in 0..EXPECTED_NUM_CORES {
+		let ready_to_run_benchmark = ready_to_run_benchmark.clone();
+
+		let handle = std::thread::spawn(move || {
```

Comment:
We should pin each thread to a single core to make the benchmark more consistent. (Start at the second core to leave the first core for the benchmark harness, and wrap around if there are less than `EXPECTED_NUM_CORES + 1`.)

Reply:
Pinning did not have the expected results it turns out on a heteregenous system like the system I have available the linux scheduler is actually better at picking the right cores to execute this threads than we can be with naive strategies.

On my Intel system I've got 8 performant cores  and 8 efficient ones, with the 8 performant ones having hyper threading enabled, so linux will report 24 cores in the sched affinity mask, the best strategy would be to entirely avoid the efficient ones and to pin the 8 threads in such a way that they are not hyper-siblings, the results are similar with what I would get if I let the linux scheduler do its thing.

So pining will become easily complicated and might backfire, hence I don't think it will reflect the real world behaviour where we don't do any pinning of the polkadot threads.

With that in mind and the fact that @ggwpez is also obtaining consistent results here: https://github.com/paritytech/polkadot-sdk/pull/5127#issuecomment-2250024809, I tend to think that we shouldn't  do any pining, what do you think @koute ?

Reply:
Right, okay, fair enough, so in that case we should start at the first core instead. AFAIK on Linux you should always get physical cores first and hyperthreaded cores last, and I'm guessing on Intel hardware P cores should also come first, so e.g. if you have 8 P cores and you'll pin to the first 8 cores in the affinity mask then that should always put you on the cores we want, no complex logic needed.

And when comparing consistency (pinning vs no pinning) you probably want to try two different scenarios:

1) load your system with something in the background (essentially simulate a noisy cloud system),
2) disable some of your CPU cores (Run `echo 0 > /sys/devices/system/cpu/cpu<number>/online` as root to disable a hardware thread, you can `cat /proc/meminfo` to see the topology)

Can you check if pinning makes a difference in these two cases? If there won't be much difference then I'm fine with not having any pinning.

Reply:
> AFAIK on Linux you should always get physical cores first and hyperthreaded cores last

Actually, the  2 hyper threads sharing the same physical core are one after the other e.g on my machine with 8 P cores and 8 E, the cat /proc/cpuinfo:

```
processor       : 0
core id         : 0
processor       : 1
core id         : 0
processor       : 2
core id         : 4
processor       : 3
core id         : 4
.......  cut
processor       : 14
core id         : 28
processor       : 15
core id         : 28
### The slow cores start here and they don't have hyperthreading.
processor       : 16
core id         : 32
processor       : 17
.......  cut
core id         : 37
processor       : 22
core id         : 38
processor       : 23
core id         : 39
```

> you'll pin to the first 8 cores in the affinity mask then that should always put you on the cores we want, no complex logic needed.

It gets un-necessary complicated because hyperthread cores are intermingled with the main ones, so you need to start parsing this information about the CPU topology and try to do you best to place the thread on the most performant cores.

> load your system with something in the background (essentially simulate a noisy cloud system)

Not sure I understand how pinning would help with that ? In a cloud environment the multi-tenancy happen at a VM or a container level, and we are doing the pinning in guest, so if we are sharing the phsical hardware with someone else then you are bound to get bad results with or without pining.

Reply:
> Actually, the 2 hyper threads share the same physical core are one after the other e.g on my machine with 8 P cores and 8 E, the cat /proc/cpuinfo:

Okay, it looks like on some CPUs it does put the hyperthreads immediately after the physical cores.

Fair enough. In which case let's leave it as-is with no pinning.

> Not sure I understand how pinning would help with that ?

A pinned thread won't be migrated to another core. If a system is not loaded then Linux will in general keep a given thread running on the same core anyway (which is why pinning on an unloaded system doesn't make much difference), but if there's a high load then that might not be true. Of course this would only help within the same VM; if the noise comes from *outside* of the VM then obviously we can't do anything about that.